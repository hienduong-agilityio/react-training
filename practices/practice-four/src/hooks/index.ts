// Index Hook files
export * from './useClickOutside';
export * from './useProject';
export * from './useSearchProject';
