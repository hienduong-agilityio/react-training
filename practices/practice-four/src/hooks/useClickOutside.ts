import { useCallback } from 'react';

/**
 * React 19 ref-style useClickOutside
 * Returns a ref callback that handles click outside behavior and cleans up.
 */
export const useClickOutside = (callback: () => void) => {
  const refCallback = useCallback(
    (node: HTMLElement | null) => {
      if (!node) return;

      const handleClick = (event: MouseEvent) => {
        if (!node.contains(event.target as Node)) {
          callback();
        }
      };

      document.addEventListener('mousedown', handleClick);

      // Cleanup function for ref
      return () => {
        document.removeEventListener('mousedown', handleClick);
      };
    },

    [callback]
  );

  return refCallback;
};
