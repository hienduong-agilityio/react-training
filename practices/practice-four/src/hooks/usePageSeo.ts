import { useEffect } from 'react';

const usePageSeo = ({
  title,
  description,
  keywords,
  ogTitle,
  ogDescription,
  ogImage,
  ogUrl
}: {
  title: string;
  description: string;
  keywords?: string;
  ogTitle?: string;
  ogDescription?: string;
  ogImage?: string;
  ogUrl?: string;
}) => {
  useEffect(() => {
    document.title = title;

    setMetaTags('name', 'description', description);
    setMetaTags('name', 'keywords', keywords ?? '');
    setMetaTags('property', 'og:title', ogTitle ?? title);
    setMetaTags('property', 'og:description', ogDescription ?? description);
    setMetaTags('property', 'og:image', ogImage ?? '');
    setMetaTags('property', 'og:url', ogUrl ?? window.location.href);

    return () => {
      // Clean up meta tags on component unmount
    };
  }, [title, description, keywords, ogTitle, ogDescription, ogImage, ogUrl]);

  const setMetaTags = (attr: string, key: string, content: string) => {
    if (content) {
      let element = document.querySelector(`meta[${attr}="${key}"]`);

      if (!element) {
        element = document.createElement('meta');
        element.setAttribute(attr, key);
        document.head.appendChild(element);
      }
      element.setAttribute('content', content);
    }
  };
};

export { usePageSeo };
