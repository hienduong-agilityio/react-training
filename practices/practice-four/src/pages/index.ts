// Index Page files
export { default as NotFoundPage } from './NotFound';
export { default as ProjectDetail } from './ProjectDetail';
export { default as ProjectFormPage } from './ProjectForm';
export { default as ProjectPage } from './ProjectTable';
