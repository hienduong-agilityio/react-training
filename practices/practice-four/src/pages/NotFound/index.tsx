// Libraries
import { Link } from 'react-router-dom';
import { usePageSeo } from '@/hooks';

// Constants
import { ROUTE } from '@/constants';

/**
 * NotFoundPage
 *
 * @returns {JSX.Element} - The NotFound element
 */
const NotFoundPage = (): JSX.Element => {
  usePageSeo({
    title: 'Page Not Found',
    description: 'The page you are looking for does not exist.',
    keywords: '404, not found, error',
    ogTitle: 'Page Not Found',
    ogDescription: 'The page you are looking for does not exist.'
  });

  return (
    <div className='flex flex-col h-screen justify-center items-center bg-gray-300'>
      <p className='text-[120px] font-extrabold text-gray-700'>404</p>
      <p className='text-2xl font-medium text-gray-600 mb-6'>Not Found</p>
      <Link
        to={ROUTE.ROOT}
        className='px-4 py-2 font-medium text-white bg-primary-500 rounded-md hover:bg-primary-600 transition-all duration-200 ease-in-out'
      >
        Go Home
      </Link>
    </div>
  );
};

export default NotFoundPage;
