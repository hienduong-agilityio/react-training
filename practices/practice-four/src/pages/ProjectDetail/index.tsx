// Libraries
import { Navigate, useParams } from 'react-router-dom';

// Components
import { ProjectDetailSkeleton } from '@/components';
import { ProjectDetailInfo } from '@/pages/ProjectDetail/ProjectDetailInfo';

// Hooks
import { usePageSeo, useProject } from '@/hooks';

// Interfaces
import { IProjectItemProps } from '@/interfaces';

const ProjectDetail = (): JSX.Element => {
  const { id } = useParams<{ id: string }>();
  const project = useProject({ id: id ?? '' });
  const { data: projectDetail, isQueryProjectDetailPending, error: projectDetailsError } = project;

  // Extract SEO data if projectDetail is valid
  const seoData =
    !isQueryProjectDetailPending && projectDetail && !Array.isArray(projectDetail)
      ? {
          title: `${projectDetail.projectName}`,
          description: projectDetail.projectName || `Details about project ${projectDetail.projectName}`,
          ogTitle: `${projectDetail.projectName}`,
          ogDescription: projectDetail.projectName || ''
        }
      : null;

  // Always call usePageSeo
  usePageSeo(
    seoData || {
      title: 'Project Details',
      description: 'A powerful tool for managing and tracking projects efficiently.'
    }
  );

  // Handle loading
  if (isQueryProjectDetailPending) return <ProjectDetailSkeleton />;

  // Handle not found
  if (!projectDetail) return <Navigate to='/404' replace />;

  // Handle error
  if (projectDetailsError) {
    return <div>Error loading project: {projectDetailsError.message}</div>;
  }

  const { manager: { managerName = 'Unknown', managerImage = '' } = {} } = projectDetail as IProjectItemProps;

  return <ProjectDetailInfo managerName={managerName} managerAvatar={managerImage} {...projectDetail} />;
};

export default ProjectDetail;
