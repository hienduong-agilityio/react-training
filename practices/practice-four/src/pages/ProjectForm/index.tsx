'use client';

// Libraries
import { useFormState } from 'react-dom';
import { Navigate, useNavigate } from 'react-router-dom';

// Hooks
import { initialProjectFormState, useProjectForm } from '@/hooks';

// Components
import { FormLayout, ProjectFormFields } from '@/components';

const ProjectFormPage = ({ title = '' }: { title: string }) => {
  const navigate = useNavigate();

  const { formAction, initialFormValues, isQueryProjectDetailPending, projectDetailsError } = useProjectForm(title);

  /**
   * useFormState is a hook that allows you to manage the state of a form.
   */
  const [formState, formActionDispatcher] = useFormState(formAction, initialProjectFormState);

  if (projectDetailsError) return <Navigate to='/404' replace />;

  const handleCancelForm = () => {
    if (window.history.length > 1) {
      navigate(-1);
    }
  };
  return (
    <FormLayout title={title}>
      <ProjectFormFields
        initialFormValues={initialFormValues}
        formErrorsMessages={formState.errors}
        isQueryProjectDetailPending={isQueryProjectDetailPending}
        title={title}
        handleCancelForm={handleCancelForm}
        formAction={formActionDispatcher}
      />
    </FormLayout>
  );
};

export default ProjectFormPage;
