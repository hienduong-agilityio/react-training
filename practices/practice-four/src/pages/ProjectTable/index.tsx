import { Suspense, useDeferredValue, useOptimistic, startTransition, useActionState, useState } from 'react';
import { useNavigate, useLocation, useSearchParams, Outlet } from 'react-router-dom';
import { useQuery, keepPreviousData } from '@tanstack/react-query';
import { useScan, Options } from 'react-scan';

// Components
import { ProjectSearch, Button, Spinner, ProjectTableManager } from '@/components';

// Constants
import { ROUTE, ROWS_PER_PAGE_OPTIONS } from '@/constants';

// Hooks
import { usePageSeo, useProject, useSearchProject } from '@/hooks';

// Services
import { getProjects } from '@/services';

// Types
import type { IProjectsQueryResult } from '@/interfaces';

const ProjectPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const [searchParams] = useSearchParams();

  usePageSeo({
    title: 'Project Management',
    description: '"A powerful tool for managing and tracking projects efficiently.',
    keywords: 'project, management, tool',
    ogTitle: 'Project Management',
    ogDescription: '"A powerful tool for managing and tracking projects efficiently.',
    ogImage: 'https://example.com/og-image.jpg',
    ogUrl: 'https://project-managerment-table.netlify.app/'
  });

  const isChildRoute = location.state?.from;

  // Get values from URL
  const currentPage = parseInt(searchParams.get('page') ?? '1');
  const rowsPerPage = parseInt(searchParams.get('rowsPerPage') ?? '10');
  const initialField = searchParams.get('searchField') ?? 'projectName';
  const initialKeyword = searchParams.get(initialField) ?? '';

  // Search state hook
  const { searchField, searchKeyword, updateSearchTerm, updateSearchField } = useSearchProject(
    initialField,
    initialKeyword
  );

  const searchAction = async (_prevState: { searchValue: string }, formData: FormData) => {
    const keyword = formData.get('search') as string;

    updateSearchTerm(keyword);

    return { searchValue: keyword };
  };

  const searchFormStateTuple = useActionState(searchAction, {
    searchValue: initialKeyword
  });

  const formActionDispatcher = searchFormStateTuple[1];

  const [optimisticState, updateOptimistic] = useOptimistic({ currentPage, rowsPerPage }, (prev, next) => ({
    ...(prev ?? {}),
    ...(next ?? {})
  }));

  const deferredKeyword = useDeferredValue(searchKeyword);

  const {
    data: projectData,
    isQueryProjectsPending,
    error
  } = useProject({
    page: optimisticState.currentPage,
    rowsPerPage: optimisticState.rowsPerPage,
    filter: { [searchField]: deferredKeyword }
  });

  const allProjectsQuery: IProjectsQueryResult = useQuery({
    queryKey: ['projects', searchField, deferredKeyword],
    queryFn: () => getProjects({ filter: { [searchField]: deferredKeyword } }),
    placeholderData: keepPreviousData
  });

  const totalPages = Math.ceil((allProjectsQuery.data?.length ?? 1) / optimisticState.rowsPerPage);

  useScan({
    enabled: import.meta.env.DEV,
    name: 'ProjectPage',
    logProps: true,
    includeComponents: ['ProjectSearch', 'ProjectTableManager', 'Button'],
    rowsPerPage: optimisticState.rowsPerPage
  } as Options & { rowsPerPage: number });

  // Pagination handlers
  const handleRowsPerPageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newRowsPerPage = parseInt(event.target.value, 10);

    startTransition(() => {
      updateOptimistic({ rowsPerPage: newRowsPerPage, currentPage: 1 });
    });

    const newParams = new URLSearchParams(searchParams.toString());

    newParams.set('rowsPerPage', newRowsPerPage.toString());
    newParams.set('page', '1');

    navigate(`?${newParams.toString()}`);
  };

  const handlePageChange = (newPage: number) => {
    startTransition(() => {
      updateOptimistic({ currentPage: newPage });
    });

    const newParams = new URLSearchParams(searchParams.toString());
    newParams.set('page', newPage.toString());

    navigate(`?${newParams.toString()}`);
  };

  // Delete modal state
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [selectedProjectId, setSelectedProjectId] = useState('');

  const handleOpenDeleteProjectModal = (id: string) => {
    startTransition(() => {
      setSelectedProjectId(id);
      setIsDeleteModalOpen(true);
    });
  };

  const handleDeleteModalClose = () => {
    startTransition(() => {
      setIsDeleteModalOpen(false);
    });
  };

  // Option builder
  const rowPerPageOption = ROWS_PER_PAGE_OPTIONS.map((option) => (
    <option key={option.value} value={option.value}>
      {option.label}
    </option>
  ));

  const handleNewProjectClick = () => {
    navigate(ROUTE.ADD_PROJECT, { state: { from: location.pathname + location.search } });
  };

  const isPending = isQueryProjectsPending || allProjectsQuery.isFetching;

  if (error ?? allProjectsQuery.error) return <div>Error loading projects</div>;

  return (
    <div className='h-full'>
      {!isChildRoute ? (
        <>
          <div className='flex justify-between pt-5 pb-7 px-3'>
            <ProjectSearch
              searchField={searchField}
              updateSearchField={updateSearchField}
              initialKeyword={initialKeyword}
              formAction={formActionDispatcher}
            />
            <Button onClick={handleNewProjectClick}>New Project</Button>
          </div>

          <Suspense fallback={<Spinner />}>
            <ProjectTableManager
              isPending={isPending}
              allProjectsQuery={allProjectsQuery}
              projectData={Array.isArray(projectData) ? projectData : []}
              rowsPerPage={optimisticState.rowsPerPage}
              totalPages={totalPages}
              currentPage={optimisticState.currentPage}
              selectedProjectId={selectedProjectId}
              isDeleteModalOpen={isDeleteModalOpen}
              rowPerPageOption={rowPerPageOption}
              handleRowsPerPageChange={handleRowsPerPageChange}
              handlePageChange={handlePageChange}
              handleOpenDeleteProjectModal={handleOpenDeleteProjectModal}
              handleDeleteModalClose={handleDeleteModalClose}
            />
          </Suspense>
        </>
      ) : (
        <Outlet />
      )}
    </div>
  );
};

export default ProjectPage;
