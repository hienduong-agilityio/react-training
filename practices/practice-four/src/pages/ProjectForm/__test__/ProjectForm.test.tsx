// Libraries
import { render, screen } from '@testing-library/react';

// Pages
import { ProjectFormPage } from '@/pages';

// Hooks
import { useProject } from '@/hooks';

jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockReturnValue({ id: '1' }),
  useNavigate: jest.fn()
}));

jest.mock('@/hooks/useProject');

jest.mock('@/stores/toastStore', () => ({
  ToastStore: jest.fn().mockReturnValue({
    showToast: jest.fn()
  })
}));

describe('ProjectFormPage', () => {
  beforeEach(() => {
    (useProject as jest.Mock).mockReturnValue({
      data: {
        id: '1',
        projectName: 'Test Project',
        manager: { managerName: 'Test Manager', managerImage: '' },
        resources: '5',
        timeline: { timeStart: '2023-01-01', timeEnd: '2023-12-31' },
        budget: 1000,
        status: 'On Track'
      },
      isQueryProjectDetailPending: false,
      isMutating: false,
      mutate: { addProject: jest.fn(), editProject: jest.fn() }
    });
  });

  it('should render ProjectFormFields and match the snapshot', () => {
    const { asFragment } = render(<ProjectFormPage title='Edit Project' />);

    const projectNameInput = screen.getByLabelText(/Project Name/i);
    expect(projectNameInput).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
