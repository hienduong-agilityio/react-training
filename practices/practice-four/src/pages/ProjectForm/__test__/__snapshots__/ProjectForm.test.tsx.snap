// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectFormPage should render ProjectFormFields and match the snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class="flex rounded-t-xl items-center bg-white justify-between p-4"
    >
      <p
        class="text-xl font-semibold text-gray-900"
      >
        Edit Project
      </p>
    </div>
    <form
      action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
      class="relative space-y-4"
    >
      <div
        class="flex flex-col gap-7 px-4"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="mb-2 capitalize text-lg font-medium text-gray-700"
            for="projectName"
          >
            Project Name *
          </label>
          <input
            class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
            id="projectName"
            name="projectName"
            pattern="[a-zA-Z0-9 ]*"
            required=""
            value="Test Project"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="mb-2 capitalize text-lg font-medium text-gray-700"
            for="managerName"
          >
            Project Manager (PM)
          </label>
          <input
            class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
            id="managerName"
            name="managerName"
            pattern="[a-zA-Z ]*"
            required=""
            value="Test Manager"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="mb-2 capitalize text-lg font-medium text-gray-700"
            for="managerImage"
          >
            Manager Image URL
          </label>
          <input
            class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
            id="managerImage"
            name="managerImage"
            pattern="^(https?:\\\\/\\\\/)?([\\\\da-z\\\\.-]+)\\\\.([a-z\\\\.]{2,6})([\\\\/\\\\w \\\\.-]*)*\\\\/?$"
            type="url"
            value=""
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="mb-2 capitalize text-lg font-medium text-gray-700"
            for="status"
          >
            Status
          </label>
          <select
            class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow text-sm h-12"
            id="status"
            name="status"
            required=""
          >
            <option
              value="On track"
            >
              On track
            </option>
            <option
              value="Potential risk"
            >
              Potential risk
            </option>
            <option
              value="At risk"
            >
              At risk
            </option>
            <option
              value="On hold"
            >
              On hold
            </option>
          </select>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="mb-2 capitalize text-lg font-medium text-gray-700"
            for="resources"
          >
            Resources
          </label>
          <input
            class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
            id="resources"
            name="resources"
            type="number"
            value="5"
          />
        </div>
        <div
          class="flex flex-col gap-4 w-1/2"
        >
          <h4
            class="text-lg font-medium text-gray-700"
          >
            Project Timeline
          </h4>
          <div
            class="flex gap-4"
          >
            <div
              class="flex flex-col w-full"
            >
              <label
                for="timeStart"
              >
                From
              </label>
              <div
                class="flex flex-col"
              >
                <input
                  class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
                  id="timeStart"
                  name="timeStart"
                  type="date"
                  value="2023-01-01"
                />
              </div>
            </div>
            <img
              alt="Arrow Right"
              class="w-5 h-5 mt-6"
            />
            <div
              class="flex flex-col w-full"
            >
              <label
                for="timeEnd"
              >
                To
              </label>
              <div
                class="flex flex-col"
              >
                <input
                  class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
                  id="timeEnd"
                  name="timeEnd"
                  type="date"
                  value="2023-12-31"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="mb-2 capitalize text-lg font-medium text-gray-700"
            for="budget"
          >
            Budget
          </label>
          <input
            class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-12"
            id="budget"
            max="1000000"
            min="0"
            name="budget"
            required=""
            type="number"
            value="1000"
          />
        </div>
      </div>
      <div
        class="flex justify-end gap-4 bg-white py-5 px-4 rounded-b-xl border-t"
      >
        <button
          class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 bg-white border-2 text-black px-6 py-3 text-lg bg-danger-400 border hover:bg-danger-500 focus:bg-danger-400 hover:text-light focus:text-light"
          type="button"
        >
          Cancel
        </button>
        <button
          class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 border-0 text-white px-6 py-3 text-lg bg-primary-500 hover:bg-primary-600 focus:bg-primary-500"
          type="submit"
        >
          Edit Project
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
